analysis/tbcshaman/src/modules/features/TotemTracker.tsx

```diff
@@ -0,0 +1,228 @@
+  summonedAt: number;
+  dismissedAt?: number;
+  dismissReason?: string;
+  duration?: number;
+  targetID?: number;
+  damageDone: number;
+  healingDone: number;
+  manaRestored: number;
+  spellsGrounded?: any;
+}
+
+class TotemTracker extends Analyzer {
+  static dependencies = {
+    combatants: Combatants,
+  };
+
+  protected combatants!: Combatants;
+
+  public totemElementEvents: TotemEventTracker = {
+    [TotemElements.Fire]: [],
+    [TotemElements.Water]: [],
+    [TotemElements.Earth]: [],
+    [TotemElements.Air]: [],
+  };
+
+  public totemEvents(totemSpellId: number) {
+    const events: TotemEvent[] = [];
+    const totemElement = GetTotemElement(totemSpellId);
+    if (totemElement === null) {
+      return events;
+    }
+    return this.totemElementEvents[totemElement].filter(
+      (e: TotemEvent) => e.totemSpellId === totemSpellId,
+    );
+  }
+
+  activeTotem(element: TotemElements) {
+    if (this.totemElementEvents[element].length === 0) {
+      return null;
+    }
+    const lastTotemSummoned = this.totemElementEvents[element][
+      this.totemElementEvents[element].length - 1
+    ];
+    if (lastTotemSummoned.dismissedAt) {
+      return null;
+    }
+    return lastTotemSummoned;
+  }
+
+  totalTotemUptime(totemIdOrElement: TotemElements | number) {
+    if (Number.isInteger(totemIdOrElement)) {
+      const totemId = totemIdOrElement as number;
+      if (this.totemEvents(totemId).length === 0) {
+        return 0;
+      }
+      return this.totemEvents(totemId)
+        .map((event) => event.duration || 0)
+        .reduce((a: number, b: number) => a + b);
+    }
+    const totemElement = totemIdOrElement as TotemElements;
+    if (this.totemElementEvents[totemElement].length === 0) {
+      return 0;
+    }
+    return this.totemElementEvents[totemElement]
+      .map((event) => event.duration || 0)
+      .reduce((a: number, b: number) => a + b);
+  }
+
+  totemUptimePercentage(totemIdOrElement: TotemElements | number): number {
+    return this.totalTotemUptime(totemIdOrElement) / this.owner.fightDuration;
+  }
+
+  // Duration is hard to get perfect, but we can do a few things to make the number we get not look so outlandish.
+  markTotemAsDismissed(element: TotemElements, timestamp: number, reason = '') {
+    if (this.totemElementEvents[element].length === 0) {
+      return;
+    }
+    if (this.totemElementEvents[element][this.totemElementEvents[element].length - 1].dismissedAt) {
+      return;
+    }
+
+    const totemEvent: TotemEvent = this.totemElementEvents[element][
+      this.totemElementEvents[element].length - 1
+    ];
+    const possibleDuration: number = timestamp - totemEvent.summonedAt;
+    const maxDuration: number = (TotemDurations as any)[totemEvent.totemSpellId] as number;
+    const duration = Math.min(possibleDuration, maxDuration);
+    totemEvent.dismissedAt = timestamp;
+    totemEvent.dismissReason = reason;
+    totemEvent.duration = duration;    
```

Comment:
```suggestion
    totemEvent.duration = duration;
```

looks like prettier is catching some trailing spaces. not sure if thats from me or not, rip.